### PR TITLE
Feature - Exception handler

### PR DESCRIPTION
**Pull request by khaitruong922**
Raise taipy.core exceptions whenever an entity or config not found. Add error handler to handle each exception.

### CODE REVIEW

File: src/taipy/rest/api/error_handler.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    return jsonify(e.messages), 400
+
+
+@blueprint.errorhandler(ConfigIdMissingException)
+def handle_config_id_missing_exception(e):
+    return jsonify({"message": e.message}), 400
+
+
+@blueprint.errorhandler(NonExistingDataNode)
+def handle_data_node_not_found(e):
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/error_handler.py:52]: Hum. 
When I see a piece of code used more than three times, I tend to make that a function.

What about a:
```
def _create_404(e: Exception) -> Tuple
  return jsonify({"message": e.message}), 404
```

then, ten times, you would `return _create_404(e)`

File: src/taipy/rest/api/resources/cycle.py [hunk 2]
```diff
@@ -25,6 +26,15 @@
 REPOSITORY = "cycle"
 
 
+def get_or_raise(cycle_id: str) -> None:
+    """Raise an exception if the data node does not exist."""
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/cycle.py]: So I don't understand the 'return' statement.

I suspect the function signature and doc are wrong.

Do you really plan to publish this function anyway?

File: src/taipy/rest/api/resources/datanode.py [hunk 3]
```diff
@@ -48,6 +49,15 @@
 REPOSITORY = "data"
 
 
+def get_or_raise(data_node_id: str) -> None:
+    """Raise an exception if the data node does not exist."""
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/datanode.py]: Same here: this function MAY return something valid.
And it's probably a private function. (-> `_get(id: str) -> DataNode`)

File: src/taipy/rest/api/resources/datanode.py [hunk 4]
```diff
@@ -125,20 +135,15 @@ def __init__(self, **kwargs):
...
-        try:
-            manager = _DataManagerFactory._build_manager()
-            manager._delete(datanode_id)
-        except NonExistingDataNode:
-            return make_response(jsonify({"message": f"Data node {datanode_id} not found"}), 404)
-        return {"msg": f"Data node {datanode_id} deleted."}
+        get_or_raise(datanode_id)
+        manager = _DataManagerFactory._build_manager()
+        manager._delete(datanode_id)
+        return {"message": f"Data node {datanode_id} deleted."}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/datanode.py]: "Data node {datanode_id} deleted."
->
"Data node {datanode_id} was deleted."

File: src/taipy/rest/api/resources/job.py [hunk 5]
```diff
@@ -103,20 +113,15 @@ def __init__(self, **kwargs):
...
     @_middleware
     def delete(self, job_id):
         manager = _JobManagerFactory._build_manager()
-        job = manager._get(job_id)
-        if not job:
-            return make_response(jsonify({"message": f"Job {job_id} not found"}), 404)
+        job = get_or_raise(job_id)
         manager._delete(job)
-        return {"msg": f"Job {job_id} deleted."}
+        return {"message": f"Job {job_id} deleted."}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/job.py]: was deleted

File: src/taipy/rest/api/resources/job.py [hunk 6]
```diff
@@ -205,29 +213,20 @@ def post(self):
...
         job = self.__create_job_from_schema(task_config_id)
-
-        if not job:
-            return {"msg": f"Task with config_id {task_config_id} not found"}, 404
-
         manager._set(job)
-
         return {
-            "msg": "Job created.",
+            "message": "Job created.",
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/job.py]: was created

File: src/taipy/rest/api/resources/pipeline.py [hunk 7]
```diff
@@ -207,18 +217,15 @@ def post(self):
...
+        pipeline = manager._get_or_create(config)
 
-            return {
-                "msg": "Pipeline created.",
-                "pipeline": response_schema.dump(_to_model(REPOSITORY, pipeline)),
-            }, 201
-        except KeyError:
-            return {"msg": f"Config id {config_id} not found"}, 404
+        return {
+            "message": "Pipeline created.",
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/pipeline.py]: was created

File: src/taipy/rest/api/resources/scenario.py [hunk 8]
```diff
@@ -103,20 +113,15 @@ def __init__(self, **kwargs):
...
     @_middleware
     def delete(self, scenario_id):
         manager = _ScenarioManagerFactory._build_manager()
-        scenario = manager._get(scenario_id)
-        if not scenario:
-            return make_response(jsonify({"message": f"Scenario {scenario_id} not found"}), 404)
+        get_or_raise(scenario_id)
         manager._delete(scenario_id)
-        return {"msg": f"Scenario {scenario_id} deleted."}
+        return {"message": f"Scenario {scenario_id} deleted."}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/scenario.py]: was

File: src/taipy/rest/api/resources/scenario.py [hunk 9]
```diff
@@ -207,18 +215,15 @@ def post(self):
...
+        scenario = manager._create(config)
 
-            return {
-                "msg": "Scenario created.",
-                "scenario": response_schema.dump(_to_model(REPOSITORY, scenario)),
-            }, 201
-        except KeyError:
-            return {"msg": f"Config id {config_id} not found"}, 404
+        return {
+            "message": "Scenario created.",
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/scenario.py]: was

File: src/taipy/rest/api/resources/task.py [hunk 10]
```diff
@@ -102,20 +113,15 @@ def __init__(self, **kwargs):
...
     @_middleware
     def delete(self, task_id):
         manager = _TaskManagerFactory._build_manager()
-        task = manager._get(task_id)
-        if not task:
-            return make_response(jsonify({"message": f"Task {task_id} not found"}), 404)
+        get_or_raise(task_id)
         manager._delete(task_id)
-        return {"msg": f"Task {task_id} deleted."}
+        return {"message": f"Task {task_id} deleted."}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/task.py]: was

File: src/taipy/rest/api/resources/task.py [hunk 11]
```diff
@@ -205,18 +214,15 @@ def post(self):
...
+        task = manager._bulk_get_or_create([config])[0]
 
-            return {
-                "msg": "Task created.",
-                "task": schema.dump(_to_model(REPOSITORY, task)),
-            }, 201
-        except KeyError:
-            return {"msg": f"Config id {config_id} not found"}, 404
+        return {
+            "message": "Task created.",
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/task.py]: was

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Please elaborate on these get_or_raise.
I suspect a local, private _get() would be enough everywhere.

File: src/taipy/rest/api/resources/cycle.py [hunk 12]
```diff
@@ -351,7 +355,7 @@ class CycleList(Resource):
                 ```
                 <Response [201]>
                 {
-                    'msg': 'cycle created.',
+                    'message': 'cycle was created.',
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/cycle.py]: Capitalize sentence.

File: src/taipy/rest/api/resources/job.py [hunk 13]
```diff
@@ -277,8 +276,6 @@ def __init__(self, **kwargs):
     @_middleware
     def post(self, job_id):
         manager = _JobManagerFactory._build_manager()
-        job = manager._get(job_id)
-        if not job:
-            return make_response(jsonify({"message": f"Job {job_id} not found"}), 404)
+        job = _get_or_raise(job_id)
         manager._cancel(job)
         return {"message": f"Cancelled job {job_id}"}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/job.py]: Check if this should be worded just like the other message: (only one 'l' by the way)

     Job {} was canceled.

File: src/taipy/rest/api/resources/pipeline.py [hunk 14]
```diff
@@ -266,9 +270,7 @@ def __init__(self, **kwargs):
...
-        try:
-            manager = _PipelineManagerFactory._build_manager()
-            manager._submit(pipeline_id)
-            return {"message": f"Executed pipeline {pipeline_id}"}
-        except NonExistingPipeline:
-            return make_response(jsonify({"message": f"Pipeline {pipeline_id} not found"}), 404)
+        _get_or_raise(pipeline_id)
+        manager = _PipelineManagerFactory._build_manager()
+        manager._submit(pipeline_id)
+        return {"message": f"Executed pipeline {pipeline_id}"}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/pipeline.py]: Same here:

Pipeline {} was executed.

(or submitted? what is the difference? asynchronism I suspect)

File: src/taipy/rest/api/resources/scenario.py [hunk 15]
```diff
@@ -266,9 +269,7 @@ def __init__(self, **kwargs):
...
-        try:
-            manager = _ScenarioManagerFactory._build_manager()
-            manager._submit(scenario_id)
-            return {"message": f"Executed scenario {scenario_id}"}
-        except NonExistingScenario:
-            return make_response(jsonify({"message": f"Scenario {scenario_id} not found"}), 404)
+        _get_or_raise(scenario_id)
+        manager = _ScenarioManagerFactory._build_manager()
+        manager._submit(scenario_id)
+        return {"message": f"Executed scenario {scenario_id}"}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/scenario.py]: See similar comment on Pipelines...

File: src/taipy/rest/api/resources/task.py [hunk 16]
```diff
@@ -265,8 +268,6 @@ def __init__(self, **kwargs):
     @_middleware
     def post(self, task_id):
         manager = _TaskManagerFactory._build_manager()
-        task = manager._get(task_id)
-        if not task:
-            return make_response(jsonify({"message": f"Task {task_id} not found"}), 404)
+        task = _get_or_raise(task_id)
         manager._scheduler().submit_task(task)
         return {"message": f"Executed task {task_id}"}
```
Comment (FabienLelaquais) [inline src/taipy/rest/api/resources/task.py]: Check here please

**Review by FabienLelaquais — DISMISSED** [review]
Tiny tiny comments on the form.

**Review by jrobinAV — APPROVED** [review]

**Review by jrobinAV — COMMENTED** [review]
Tests are broken

**Comment by khaitruong922** [comment]
Tests will not be broken anymore after the fix to taipy-core

**Pull request merged**